README.md
# @vivaxy/vlq

[![Build Status][travis-image]][travis-url]
[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Standard Version][standard-version-image]][standard-version-url]
[![Codecov][codecov-image]][codecov-url]

## Install

`yarn add @vivaxy/vlq` or `npm i @vivaxy/vlq`

## Usage

```js
import { encode, decode } from '@vivaxy/vlq';
// encode
encode(0);
// decode
decode('A');
```

## API

`encode(input: number | number[]): string`

`decode(input: string): number[]`

## Benchmark

### decode

vlq#decode x 1,103,260 ops/sec ±0.62% (91 runs sampled)

@vivaxy/vlq#decode x 1,935,656 ops/sec ±0.47% (95 runs sampled)

Fastest is @vivaxy/vlq#decode

### encode

vlq#encode x 4,445,119 ops/sec ±0.91% (91 runs sampled)

@vivaxy/vlq#encode x 3,454,481 ops/sec ±0.76% (92 runs sampled)

Fastest is vlq#encode

## Related Projects

- [vlq](https://github.com/Rich-Harris/vlq)

_(Project created by [create-n](https://github.com/vivaxy/create-n).)_

[travis-image]: https://img.shields.io/travis/vivaxy/vlq.svg?style=flat-square
[travis-url]: https://travis-ci.org/vivaxy/vlq
[npm-version-image]: https://img.shields.io/npm/v/@vivaxy/vlq.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@vivaxy/vlq
[npm-downloads-image]: https://img.shields.io/npm/dt/@vivaxy/vlq.svg?style=flat-square
[license-image]: https://img.shields.io/npm/l/@vivaxy/vlq.svg?style=flat-square
[license-url]: LICENSE
[standard-version-image]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg?style=flat-square
[standard-version-url]: https://github.com/conventional-changelog/standard-version
[codecov-image]: https://img.shields.io/codecov/c/github/vivaxy/vlq.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/vivaxy/vlq


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2020-03-07)


### Features

* :sparkles:  Initial commit ([c89e92c](https://github.com/vivaxy/vlq/commit/c89e92cc7d5ee9d0d58c14f821c61f196fa4b7f3))


src/__tests__/index.test.ts
/**
 * @since 20180911 17:16
 * @author vivaxy
 */
import { encode, decode } from '../index';

test('encode', function() {
  expect(encode(0)).toBe('A');
  expect(encode(1)).toBe('C');
  expect(encode(-1)).toBe('D');
  expect(encode(23)).toBe('uB');
  expect(encode(-123)).toBe('3H');
  expect(encode([0])).toBe('A');
});

test('decode', function() {
  expect(decode('A')).toStrictEqual([0]);
  expect(decode('B')).toStrictEqual([-0]);
  expect(decode('C')).toStrictEqual([1]);
  expect(decode('D')).toStrictEqual([-1]);
  expect(decode('3H')).toStrictEqual([-123]);
  expect(decode('ktC')).toStrictEqual([1234]);
  expect(decode('EAEE')).toStrictEqual([2, 0, 2, 2]);
  expect(() => decode('?')).toThrow('Invalid character (?)');
});


src/index.ts
const integerToChar =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
let charToInteger: { [char: string]: number } = {};

for (let i = 0; i < integerToChar.length; i++) {
  charToInteger[integerToChar[i]] = i;
}

export function encode(input: number | number[]): string {
  if (typeof input === 'number') {
    return encodeInteger(input);
  }
  let result = '';
  for (const v of input) {
    result += encodeInteger(v);
  }
  return result;
}

function encodeInteger(input: number): string {
  let result = '';
  const sign = input < 0 ? ((input = -input), 1) : 0;
  const bits = (input & 15) << 1;
  let continuation = (input >>= 4) === 0 ? 0 : 32;
  result += integerToChar[bits | continuation | sign];
  while (continuation) {
    const bits = input & 31;
    continuation = (input >>= 5) === 0 ? 0 : 32;
    result += integerToChar[bits | continuation];
  }
  return result;
}

export function decode(input: string): number[] {
  let result = [];
  let number = 0;
  let continuation = 0;
  let sign = 0;
  let shift = 0;

  for (const char of input) {
    let integer = charToInteger[char];

    if (integer === undefined) {
      throw new Error('Invalid character (' + char + ')');
    }

    continuation = integer & 32;
    integer &= 31;
    if (shift === 0) {
      sign = integer & 1;
      integer >>= 1;
    }

    number |= integer << shift;

    if (!continuation) {
      result.push(sign ? -number : number);
      number = 0;
      continuation = 0;
      sign = 0;
      shift = 0;
      continue;
    }

    if (shift === 0) {
      shift = 4;
    } else {
      shift += 5;
    }
  }
  return result;
}


benchmark/index.js
/**
 * @since 2020-03-07 05:46
 * @author vivaxy
 */
const path = require('path');
const fse = require('fs-extra');
const Benchmark = require('benchmark');
const vlq = require('vlq');
const myVlq = require('../lib');

function benchmark(targetName, targetFunction, myName, myFunction) {
  return new Promise(function(resolve) {
    let message = '';
    const suite = new Benchmark.Suite();
    suite
      .add(targetName, targetFunction)
      .add(myName, myFunction)
      .on('cycle', function(event) {
        message += String(event.target) + '\n\n';
      })
      .on('complete', async function() {
        message += 'Fastest is ' + this.filter('fastest').map('name');
        resolve(message);
      })
      .run();
  });
}

async function writeToReadme(content) {
  const readmePath = path.join(__dirname, '..', 'README.md');
  const splitString = '\n## ';
  const readme = await fse.readFile(readmePath, 'utf8');
  const readmes = readme.split(splitString).map((readmeSection) => {
    if (readmeSection.startsWith('Benchmark')) {
      return 'Benchmark\n\n' + content + '\n';
    }
    return readmeSection;
  });
  await fse.writeFile(readmePath, readmes.join(splitString));
}

(async function() {
  const encodeCases = [1, -1, 0, 234, 56, 789, 10, [-1, 23, 456, 7890]];
  function forEachEncodeCase(fn) {
    encodeCases.forEach(fn);
  }
  const encodeResult = await benchmark(
    'vlq#encode',
    function() {
      forEachEncodeCase(vlq.encode);
    },
    '@vivaxy/vlq#encode',
    function() {
      forEachEncodeCase(myVlq.encode);
    },
  );

  const decodeCases = ['A', 'B', 'C', 'D', '3H', 'EAEE'];
  function forEachDecodeCase(fn) {
    decodeCases.forEach(fn);
  }
  const decodeResult = await benchmark(
    'vlq#decode',
    function() {
      forEachDecodeCase(vlq.encode);
    },
    '@vivaxy/vlq#decode',
    function() {
      forEachDecodeCase(myVlq.decode);
    },
  );
  const result = `### decode

${decodeResult}

### encode

${encodeResult}`;
  await writeToReadme(result);
  console.log(result);
})();
